lom/src/engine/context.go

```diff
@@ -0,0 +1,386 @@
+            }
+        }
+
+        tnow := time.Now().Unix()
+        tout = A_DAY_IN_SECS
+        for i := 0; i < len(listWTimeout); i++ {
+            if tnow >= listWTimeout[i].due {
+                /* Fail the request */
+                r := listWTimeout[i].req
+                r.ChResponse <- &LoMResponse { int(LoMReqTimeout),
+                                GetLoMResponseStr(LoMReqTimeout), nil }
+            } else {
+                /* Get tout and drop the failed ones */
+                tout = 0
+                if tnow < listWTimeout[i].due {
+                    tout = listWTimeout[i].due - tnow
+                }
+                listWTimeout = listWTimeout[i:]
+                break
+            }
+        }
+
+    }
+}
+
+
+/*
+ * A go func watches, yet it may get blocked by actions like publish
+ * To ensure, notify APIs are not blocked, have a buffer.
+ * Publish API is generally pretty quick, a value of 10 is more than enough.
+ */
+const HEARTBEAT_CH_SIZE = 10
+
+/* All registrations & heartbeats from clients. */
+type ClientRegistrations_t struct {
+    activeActions   map[string]*ActiveActionInfo_t
+    activeClients   map[string]*ActiveClientInfo_t
+    heartbeatCh     chan string     /* Action name */
+}
+
+/* Initialized object; But not exported */
+var clientRegistrations *ClientRegistrations_t = nil
+
+
+func GetRegistrations() *ClientRegistrations_t {
+    if clientRegistrations == nil {
+        InitRegistrations()
+    }
+    return clientRegistrations
+}
+
+
+func InitRegistrations() *ClientRegistrations_t {
+    clientRegistrations = &ClientRegistrations_t{
+                make(map[string]*ActiveActionInfo_t),
+                make(map[string]*ActiveClientInfo_t),
+                make(chan string, HEARTBEAT_CH_SIZE),
+            }
+    go clientRegistrations.PublishHeartbeats()
+    return clientRegistrations
+}
+
+func (p *ClientRegistrations_t) RegisterClient(name string) error {
+    if len(name) == 0 {
+        return LogError("Expect non empty name")
+    }
+    if _, ok := p.activeClients[name]; ok {
+        LogInfo("%s: Duplicate client registration; De & re-register", name)
+        p.DeregisterClient(name)
+    }
+    cl := &ActiveClientInfo_t {
+        ClientName: name,
+        Actions: make(map[string]struct{}),
+        pendingWriteRequests: make(chan *ServerRequestData, CHAN_REQ_SIZE),
+        pendingReadRequests: make(chan *LoMRequestInt, CHAN_REQ_SIZE),
+        abortCh: make(chan interface{}, 2) }
+                
+    p.activeClients[name] = cl
+    go cl.ProcessSendRequests()
+    return nil
+}
+
+func (p *ClientRegistrations_t) RegisterAction(action *ActiveActionInfo_t) error {
+    if action == nil {
+        return LogError("Expect non nil ActiveActionInfo_t")
+    }
+    cl, ok := p.activeClients[action.Client]
+    if !ok {
+        return LogError("%s: Missing client registration", action.Client)
+    } else if r, ok1 := p.activeActions[action.Action]; ok1 {
```

Comment:
Could you use more descriptive variable names ?

Reply:
s/action/actionInfo/g -- Done